### PR TITLE
[cxx-interop] Mark types with a destructor a non-trivial.

### PR DESCRIPTION
**Pull request by zoecarver**
Any C++ type that isn't trivially copyable is now also not a trivial type. This will preserve the destructor (among other things).

### CODE REVIEW

**Comment by gribozavr** [comment]
+ @gottesmm to review SILGen changes.

File: test/Interop/Cxx/class/Inputs/non-trivial.h [hunk 1]
```diff
@@ -0,0 +1,12 @@
+#ifndef TEST_INTEROP_CXX_CLASS_INPUTS_NON_TRIVIAL_H
```
[the remark this answers is not in the document]
Comment (zoecarver) [reply test/Interop/Cxx/class/Inputs/non-trivial.h:1]: SGTM. Done.

File: lib/SILGen/SILGenStmt.cpp [hunk 2]
```diff
@@ -379,18 +379,20 @@ prepareIndirectResultInit(SILGenFunction &SGF,
...
-
-    for (auto resultEltType : resultTupleType.getElementTypes()) {
-      auto eltInit = prepareIndirectResultInit(SGF, fnTypeForResults,
-                                               resultEltType, allResults,
-                                               directResults,
-                                               indirectResultAddrs, cleanups);
-      tupleInit->SubInitializations.push_back(std::move(eltInit));
-    }
+    // If this is an empty tuple, we don't want to decompose it.
+    if (resultTupleType.getElementTypes().size()) {
```
Comment (gottesmm) [inline lib/SILGen/SILGenStmt.cpp:383]: Did this need to be changed?

Comment (gottesmm) [reply lib/SILGen/SILGenStmt.cpp:383]: IOW, I don't understand why this change requires this other fix. Does a crash result? Please explain this to me?

Comment (zoecarver) [reply lib/SILGen/SILGenStmt.cpp:383]: Yes. Using any type with a user-provided destructor crashes without this fix. 

When we have an empty tuple, we get the wrong return type later on. Because the value is returned with an indirect result, the return type should be `@out Foo` but instead is `()` (because that's what's returned).

Comment (zoecarver) [reply lib/SILGen/SILGenStmt.cpp:383]: TBH, I'm not sure this is the right fix. It did fix the problem and (seems like?) it doesn't have any unwanted side effects but, it doesn't fix the "root" of the issue. I can investigate this a bit more.

Comment (gottesmm) [reply lib/SILGen/SILGenStmt.cpp:383]: Yes. Maybe @jckarter or @slavapestov have ideas.

Comment (jckarter) [reply lib/SILGen/SILGenStmt.cpp:383]: Yeah, it seems like a reasonable optimization to avoid forming a tuple initialization for an empty tuple, but it doesn't seem like it should be necessary for correctness. How does it crash without this change?

Comment (zoecarver) [reply lib/SILGen/SILGenStmt.cpp:383]: @jckarter the assertion on line 451: `assert(allResults.empty());`. 

Because in `prepareIndirectResultInit` the tuple path is taken and the tuple is empty, the results aren't recursively `slice`d off `allResults`.

Comment (zoecarver) [reply lib/SILGen/SILGenStmt.cpp:383]: FWIW I suspect this is probably slower (it goes from essentially a no-op to a recursive function that _will_ dynamically allocate memory).

Comment (jckarter) [reply lib/SILGen/SILGenStmt.cpp:383]: @zoecarver If the result type is an empty tuple, though, then there shouldn't be anything in `allResults` to slice off to begin with. Do you know what the root of the change in behavior is?

Comment (zoecarver) [reply lib/SILGen/SILGenStmt.cpp:383]: @jckarter sorry this comment slipped through my inbox somehow. I can spend some time today looking into this in more depth. 

> Do you know what the root of the change in behavior is?

I don't know.

> If the result type is an empty tuple, though, then there shouldn't be anything in allResults to slice off to begin with.

I think `allResults` holds indirect results (also?).

File: lib/ClangImporter/ImportDecl.cpp [hunk 3]
```diff
@@ -1272,11 +1272,7 @@ synthesizeStructDefaultConstructorBody(AbstractFunctionDecl *afd,
   auto assign = new (ctx) AssignExpr(lhs, SourceLoc(), call, /*implicit*/ true);
   assign->setType(emptyTuple);
 
-  auto result = TupleExpr::createEmpty(ctx, SourceLoc(), SourceLoc(),
-                                       /*Implicit=*/true);
-  result->setType(emptyTuple);
-
-  auto ret = new (ctx) ReturnStmt(SourceLoc(), result, /*Implicit=*/true);
+  auto ret = new (ctx) ReturnStmt(SourceLoc(), nullptr, /*Implicit=*/true);
```
Comment (zoecarver) [inline lib/ClangImporter/ImportDecl.cpp:1275]: I'm going to make separate PRs for this and the change to `IRGen/GenBuiltin.cpp` with tests.

File: lib/SILGen/SILGenStmt.cpp (hunk 2 above)
Comment (zoecarver) [reply lib/SILGen/SILGenStmt.cpp]: I was able to find and fix the root of the problem (the return statement was given an expression when it shouldn't have).

**Comment by zoecarver** [comment]
@gribozavr friendly ping. This PR no longer has any SILGen changes.

File: test/Interop/Cxx/class/Inputs/type-classification.h [hunk 4]
```diff
@@ -155,4 +155,12 @@ struct StructDeletedDestructor {
   ~StructDeletedDestructor() = delete;
 };
 
+struct HasDestructor {
```
Comment (gribozavr) [inline test/Interop/Cxx/class/Inputs/type-classification.h:158]: Could you use `StructWithDestructor` instead?

File: test/Interop/Cxx/class/Inputs/type-classification.h [hunk 5]
```diff
@@ -155,4 +155,12 @@ struct StructDeletedDestructor {
...
+  ~HasDestructor() {}
+};
+
+struct HasMemberWithDestructor {
```
Comment (gribozavr) [inline test/Interop/Cxx/class/Inputs/type-classification.h:162]: Could you use `StructWithSubobjectDestructor` instead?

**Review by gribozavr — APPROVED** [review]
LGTM with suggested changes.

**Comment by zoecarver** [comment]
@swift-ci please smoke test and merge

**Comment by zoecarver** [comment]
@swift-ci please smoke test and merge

File: test/Interop/Cxx/class/type-classification-non-trivial-silgen.swift [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+// "destroy_addr".
+// CHECK-LABEL: @$s4main24testStructWithDestructoryyF
+// CHECK: [[AS:%.*]] = alloc_stack $StructWithDestructor
+// CHECK: destroy_addr [[AS]]
+// CHECK-LABEL: end sil function '$s4main24testStructWithDestructoryyF'
+public func testStructWithDestructor() {
+  let d = StructWithDestructor()
+}
+
+// Make sure that "HasMemberWithDestructor" is marked as non-trivial by checking
```
Comment (gribozavr) [inline test/Interop/Cxx/class/type-classification-non-trivial-silgen.swift:15]: `HasMemberWithDestructor` => `StructWithSubobjectDestructor`?

Comment (gribozavr) [reply test/Interop/Cxx/class/type-classification-non-trivial-silgen.swift:15]: @zoecarver ^^^

Comment (zoecarver) [reply test/Interop/Cxx/class/type-classification-non-trivial-silgen.swift:15]: Oops forgot about this. #33150.
